va/caaQuorum.go

```diff
@@ -0,0 +1,246 @@
+	TTL  int    `json:"TTL"`
+	Data string `json:"data"`
+}
+
+type response struct {
+	// Ignored fields
+	//   tc
+	//   rd
+	//   ra
+	//   ad
+	//   cd
+	//   question
+	//   additional
+	//   edns_client_subnet
+	Status  int      `json:"Status"`
+	Answer  []answer `json:"Answer"`
+	Comment string   `json:"Comment"`
+}
+
+func parseAnswer(as []answer) ([]*dns.CAA, error) {
+	rrs := []*dns.CAA{}
+	// only bother parsing out CAA records
+	for _, a := range as {
+		if a.Type != 257 {
+			continue
+		}
+		rr, err := dns.NewRR(fmt.Sprintf("%s %d IN %s %s", a.Name, a.TTL, dns.TypeToString[a.Type], a.Data))
+		if err != nil {
+			return nil, err
+		}
+		if caaRR, ok := rr.(*dns.CAA); ok {
+			rrs = append(rrs, caaRR)
+		}
+	}
+	return rrs, nil
+}
+
+func createClient(timeout, keepAlive time.Duration, proxy string) (*http.Client, string, error) {
+	u, err := url.Parse(proxy)
+	if err != nil {
+		return nil, "", err
+	}
+	transport := &http.Transport{
+		Proxy: http.ProxyURL(u),
+		Dial: (&net.Dialer{
+			Timeout:   timeout,
+			KeepAlive: keepAlive,
+		}).Dial,
+		TLSHandshakeTimeout: 10 * time.Second,
+	}
+	return &http.Client{
+		Transport: transport,
+		Timeout:   timeout,
+	}, u.Host, nil
+}
+
+// CAAPublicResolver holds state needed to talk to GPDNS
+type CAAPublicResolver struct {
+	clients     map[string]*http.Client
+	stats       metrics.Scope
+	maxFailures int
+}
+
+// NewCAAPublicResolver returns a initialized CAAPublicResolver
+func NewCAAPublicResolver(scope metrics.Scope, timeout, keepAlive time.Duration, maxFailures int, proxies []string) (*CAAPublicResolver, error) {
+	cpr := &CAAPublicResolver{stats: scope, maxFailures: maxFailures}
+	for _, p := range proxies {
+		c, h, err := createClient(timeout, keepAlive, p)
+		if err != nil {
+			return nil, err
+		}
+		cpr.clients[h] = c
+	}
+	return cpr, nil
+}
+
+func (cpr *CAAPublicResolver) queryCAA(ctx context.Context, req *http.Request, ic *http.Client) ([]*dns.CAA, error) {
+	apiResp, err := ctxhttp.Do(ctx, ic, req)
+	if err != nil {
+		return nil, err
+	}
+	defer func() {
+		_ = apiResp.Body.Close()
+	}()
+	body, err := ioutil.ReadAll(apiResp.Body)
+	if err != nil {
+		return nil, err
+	}
+	var respObj response
+	err = json.Unmarshal(body, &respObj)
+	if err != nil {
+		return nil, err
+	}
+	if respObj.Status != dns.RcodeSuccess {
+		if respObj.Comment != "" {
+			return nil, fmt.Errorf("Query failed with %s: %s", dns.RcodeToString[respObj.Status], respObj.Comment)
+		}
+		return nil, fmt.Errorf("Query failed wtih %s", dns.RcodeToString[respObj.Status])
+	}
+
+	return parseAnswer(respObj.Answer)
+}
+
+type queryResult struct {
+	records []*dns.CAA
+	err     error
+}
+
+type caaSet []*dns.CAA
+
+func (cs caaSet) Len() int           { return len(cs) }
+func (cs caaSet) Less(i, j int) bool { return cs[i].Value < cs[j].Value } // sort by value...?
+func (cs caaSet) Swap(i, j int)      { cs[i], cs[j] = cs[j], cs[i] }
+
+func hashCAASet(set []*dns.CAA) ([32]byte, error) {
```

Comment:
I did. I'm happy with forgoing hashing, I don't think we can rely on either Unbound or remote resolvers sorting their answers as part of the check though (I believe you can explicitly set at least BIND to return answers in a random order).